dango/types/src/vesting.rs

```diff
@@ -0,0 +1,183 @@
+use {
+    grug::{
+        Addr, Coin, Duration, MultiplyFraction, Number, NumberConst, Timestamp, Udec128, Uint128,
+        Undefined,
+    },
+    std::{cmp::min, collections::BTreeMap},
+};
+
+pub type PositionIndex = u32;
+
+pub type ClaimablePosition = Position<Uint128>;
+
+#[grug::derive(Serde)]
+pub struct InstantiateMsg {
+    pub unlocking_cliff: Duration,
+    pub unlocking_vesting: Duration,
+}
+
+#[grug::derive(Serde)]
+pub enum ExecuteMsg {
+    /// Create a vesting position for a user with the given schedule.
+    /// Sender must attach a single coin.
+    CreatePosition {
+        user: Addr,
+        schedule: Schedule,
+    },
+    /// Claim the withdrawable amount from the vesting position.
+    Claim {
+        idx: PositionIndex,
+    },
+    // Terminate the vesting position.
+    // When terminated, the snapshot of the vested so far is taken and stored
+    TerminatePosition {
+        idx: PositionIndex,
+    },
+}
+
+#[grug::derive(Serde, QueryRequest)]
+pub enum QueryMsg {
+    /// Query a single vesting position by index.
+    #[returns(ClaimablePosition)]
+    Position { idx: PositionIndex },
+    /// Enumerate all vesting positions.
+    #[returns(BTreeMap<PositionIndex, ClaimablePosition>)]
+    Positions {
+        start_after: Option<PositionIndex>,
+        limit: Option<u32>,
+    },
+    /// Enumerate all vesting positions belonging to a given user.
+    #[returns(BTreeMap<PositionIndex, ClaimablePosition>)]
+    PositionsByUser {
+        user: Addr,
+        start_after: Option<PositionIndex>,
+        limit: Option<u32>,
+    },
+}
+
+#[grug::derive(Serde, Borsh)]
+pub struct Config {
+    pub unlocking_schedule: Schedule,
+}
+
+#[grug::derive(Serde, Borsh)]
+pub struct Schedule {
+    pub start_time: Timestamp,
+    pub cliff: Duration,
+    pub vesting: Duration,
```

Comment:
This should be named `period` instead of vesting, because the unlocking schedule also has a period...